pinot-segment-local/src/main/java/org/apache/pinot/segment/local/recordtransformer/ComplexTypeTransformer.java

```diff
@@ -0,0 +1,184 @@
+        && tableConfig.getIngestionConfig().getComplexTypeHandlingConfig() != null) {
+      _collectionsToUnnest =
+          tableConfig.getIngestionConfig().getComplexTypeHandlingConfig().getUnnestConfig() != null ? tableConfig
+              .getIngestionConfig().getComplexTypeHandlingConfig().getUnnestConfig() : new ArrayList<>();
+    } else {
+      _collectionsToUnnest = new ArrayList<>();
+    }
+  }
+
+  @VisibleForTesting
+  public ComplexTypeTransformer(List<String> unnestCollections) {
+    _collectionsToUnnest = new ArrayList<>(unnestCollections);
+    Collections.sort(_collectionsToUnnest);
+  }
+
+  public static boolean isComplexTypeHandlingEnabled(TableConfig tableConfig) {
+    if (tableConfig.getIngestionConfig() == null
+        || tableConfig.getIngestionConfig().getComplexTypeHandlingConfig() == null
+        || tableConfig.getIngestionConfig().getComplexTypeHandlingConfig().getMode() == null) {
+      return false;
+    }
+    return tableConfig.getIngestionConfig().getComplexTypeHandlingConfig().getMode()
+        != ComplexTypeHandlingConfig.Mode.NONE;
+  }
+
+  @Nullable
+  @Override
+  public GenericRow transform(GenericRow record) {
+    flattenMap(record, new HashSet<>(record.getFieldToValueMap().keySet()));
+    for (String collection : _collectionsToUnnest) {
+      unnestCollection(record, collection);
+    }
+    return record;
+  }
+
+  private GenericRow unnestCollection(GenericRow record, String column) {
+    if (record.getValue(GenericRow.MULTIPLE_RECORDS_KEY) == null) {
+      List<GenericRow> list = new ArrayList<>();
+      unnestCollection(record, column, list);
+      record.putValue(GenericRow.MULTIPLE_RECORDS_KEY, list);
+    } else {
+      Collection<GenericRow> records = (Collection) record.getValue(GenericRow.MULTIPLE_RECORDS_KEY);
+      List<GenericRow> list = new ArrayList<>();
+      for (GenericRow innerRecord : records) {
+        unnestCollection(innerRecord, column, list);
+      }
+      record.putValue(GenericRow.MULTIPLE_RECORDS_KEY, list);
+    }
+    return record;
+  }
+
+  private void unnestCollection(GenericRow record, String column, List<GenericRow> list) {
+    Object value = record.removeValue(column);
+    if (value == null) {
+      // use the record itself
+      list.add(record);
+      return;
+    } else if (value instanceof Collection) {
+      if (((Collection) value).isEmpty()) {
+        // use the record itself
+        list.add(record);
+      } else {
+        for (Object obj : (Collection) value) {
+          GenericRow copy = flattenCollectionItem(record, obj, column);
+          list.add(copy);
+        }
+      }
+    } else if (value.getClass().isArray()) {
+      if (((Object[]) value).length == 0) {
+        // use the record itself
+        list.add(record);
+      } else {
+        for (Object obj : (Object[]) value) {
+          GenericRow copy = flattenCollectionItem(record, obj, column);
+          list.add(copy);
+        }
```

Comment:
Can you give an example on how we are differentiating between Collection and Array. In the source data, anything in the square brackets is candidate for collection-unnesting rule right ? When will the GenericRow store it as a Collection v/s Array ?

I think if the [] has primitives, isArray() will be true whereas if [] has non-primitives Collection will be true ?

Reply:
Good question. I think we dont have this explicit on array vs collection, based on my read of the decoder code. Currently, we always convert collections in source format into array. It's understandable coz today we need array to store primitives. However, for objects, IMO collections are preferred. e.g. https://stackoverflow.com/questions/13361489/differences-between-an-array-and-any-collection-from-the-java-collection-framewo